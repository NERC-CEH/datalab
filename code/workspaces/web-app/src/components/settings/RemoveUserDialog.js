import React from 'react';
import Dialog from '@mui/material/Dialog';
import DialogTitle from '@mui/material/DialogTitle';
import DialogContent from '@mui/material/DialogContent';
import DialogContentText from '@mui/material/DialogContentText';
import DialogActions from '@mui/material/DialogActions';
import SecondaryActionButton from '../common/buttons/SecondaryActionButton';
import DangerButton from '../common/buttons/DangerButton';

export default function RemoveUserDialog({ classes, state, setState, onRemoveConfirmationFn, projectKey, dispatch }) {
  if (!state.open) return null;

  const closedState = { user: null, open: false };
  return (
    <Dialog open={state.open}>
      <DialogTitle>Remove User Permissions?</DialogTitle>
      <DialogContent>
        <DialogContentText>
          {`Are you sure you want to remove all project permissions for ${state.user.name}?`}
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <DangerButton
          id="remove-button"
          className={classes.dialogDeleteUserButton}
          variant="outlined"
          onClick={() => {
            onRemoveConfirmationFn(projectKey, state.user, dispatch);
            setState(closedState);
          }}
        >
          Remove
        </DangerButton>
        <SecondaryActionButton id="cancel-button" onClick={() => setState(closedState)}>
          Cancel
        </SecondaryActionButton>
      </DialogActions>
    </Dialog>
  );
}
