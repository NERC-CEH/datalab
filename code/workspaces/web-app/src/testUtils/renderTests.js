import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider as MuiThemeProvider } from '@mui/material/styles';
import theme from '../theme';

const AllTheProviders = ({ children }) => (
    <MuiThemeProvider theme={theme}>
        {children}
    </MuiThemeProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
