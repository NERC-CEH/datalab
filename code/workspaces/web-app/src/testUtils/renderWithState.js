import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { render } from '@testing-library/react';
import promise from 'redux-promise-middleware';
import {
  PROMISE_TYPE_PENDING,
  PROMISE_TYPE_SUCCESS,
  PROMISE_TYPE_FAILURE,
} from '../actions/actionTypes';

const promiseTypeSuffixes = [PROMISE_TYPE_PENDING, PROMISE_TYPE_SUCCESS, PROMISE_TYPE_FAILURE];

const TEST_USER = 'test-user';

export const buildDefaultTestState = () => ({
  authentication: {
    identity: {
      sub: TEST_USER,
    },
    permissions: {
      fetching: { inProgress: false },
      value: [],
    },
  },
  projectUsers: {
    fetching: { inProgress: false },
    value: [
      { userId: TEST_USER, role: 'admin' },
    ],
  },
});

export const renderWithState = (
  initialState,
  ComponentToRender,
  props = {},
  renderOptions = {},
) => {
  const mockStore = configureStore([promise({ promiseTypeSuffixes })]);
  const store = mockStore(initialState);

  const wrapper = render(
    <Provider store={store}>
      <ComponentToRender {...props} />
    </Provider>,
    renderOptions,
  );

  return {
    wrapper,
    store,
  };
};
