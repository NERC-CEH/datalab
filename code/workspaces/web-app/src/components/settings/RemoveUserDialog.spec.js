import React from 'react';
import { render, screen, fireEvent } from '../../testUtils/renderTests';
import RemoveUserDialog from './RemoveUserDialog';

describe('RemoveUserDialog', () => {
  const classes = { dialogDeleteUserButton: 'dialogDeleteUserButton' };

  describe('when the dialog state has open equal to false', () => {
    it('renders as null', () => {
      render(
        <RemoveUserDialog
          classes={classes}
          state={{ user: null, open: false }}
          setState={jest.fn()}
          onRemoveConfirmationFn={jest.fn()}
          dispatch={jest.fn()}
        />,
      );
      expect(screen.queryByText('Remove User Permissions?')).toBeNull();
    });
  });

  describe('when the dialog state has open equal to true and a user set', () => {
    const openState = { user: { name: 'User One', userId: 'user-one-id' }, open: true };

    it('renders to match snapshot', () => {
      render(
        <RemoveUserDialog
          classes={classes}
          state={openState}
          setState={jest.fn()}
          onRemoveConfirmationFn={jest.fn()}
          dispatch={jest.fn()}
        />,
      );

      expect(screen.getByText('Remove User Permissions?').parentElement.parentElement.parentElement).toMatchSnapshot();
    });

    it('closes when the cancel button is pressed and does not call confirmation function', () => {
      const mockSetState = jest.fn();
      const mockOnRemoveConfirmationFn = jest.fn();
      render(
        <RemoveUserDialog
          classes={classes}
          state={openState}
          setState={mockSetState}
          onRemoveConfirmationFn={mockOnRemoveConfirmationFn}
          dispatch={jest.fn()}
        />,
      );
      fireEvent.click(document.querySelector('[id="cancel-button"]'));

      expect(mockSetState).toHaveBeenCalledTimes(1);
      expect(mockSetState).toHaveBeenCalledWith({ open: false, user: null });
      expect(mockOnRemoveConfirmationFn).toHaveBeenCalledTimes(0);
    });

    it('calls provided function and closes when the confirm button is pressed', () => {
      const mockSetState = jest.fn();
      const mockOnRemoveConfirmationFn = jest.fn();
      const mockDispatch = jest.fn();
      const projectKey = 'project';
      render(
        <RemoveUserDialog
          classes={classes}
          state={openState}
          setState={mockSetState}
          onRemoveConfirmationFn={mockOnRemoveConfirmationFn}
          projectKey={projectKey}
          dispatch={mockDispatch}
        />,
      );
      fireEvent.click(document.querySelector('[id="remove-button"]'));

      expect(mockSetState).toHaveBeenCalledTimes(1);
      expect(mockSetState).toHaveBeenCalledWith({ open: false, user: null });
      expect(mockOnRemoveConfirmationFn).toHaveBeenCalledTimes(1);
      expect(mockOnRemoveConfirmationFn)
        .toHaveBeenCalledWith(projectKey, openState.user, mockDispatch);
    });
  });
});
