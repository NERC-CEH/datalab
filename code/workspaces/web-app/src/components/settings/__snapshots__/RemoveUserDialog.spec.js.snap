// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveUserDialog when the dialog state has open equal to true and a user set renders to match snapshot 1`] = `
<div
  class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-testid="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
    role="presentation"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    tabindex="-1"
  >
    <div
      aria-labelledby="mui-1"
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1s9m6is-MuiPaper-root-MuiDialog-paper"
      role="dialog"
    >
      <h2
        class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-xq3hxg-MuiTypography-root-MuiDialogTitle-root"
        id="mui-1"
      >
        Remove User Permissions?
      </h2>
      <div
        class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
      >
        <p
          class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiDialogContentText-root css-1k6zr5f-MuiTypography-root-MuiDialogContentText-root"
        >
          Are you sure you want to remove all project permissions for User One?
        </p>
      </div>
      <div
        class="MuiDialogActions-root MuiDialogActions-spacing css-knqc4i-MuiDialogActions-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium dialogDeleteUserButton css-o0h1hg-MuiButtonBase-root-MuiButton-root"
          id="remove-button"
          tabindex="0"
          type="button"
        >
          Remove
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-i65xv2-MuiButtonBase-root-MuiButton-root"
          id="cancel-button"
          tabindex="0"
          type="button"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    data-testid="sentinelEnd"
    tabindex="0"
  />
</div>
`;
